### DISCUSSION TITLE
Can you replace variable's string text with method?

### DISCUSSION THREAD

**Discussion by tilenzalar** [dotnet/roslyn#50947] [Q&A]
I have this part of code that replaces all hardcoded strings in solution that have text "Quantity" with "Quantity Roslyn". Now I want to change the code so that I replace all hardcoded strings that have text "Quantity" with a function, let's say: someFunction("something").

                foreach (var projectId in workspace.CurrentSolution.ProjectIds)
                {
                var project = solution.GetProject(projectId);
                foreach (var documentId in project.DocumentIds)
                {
                    Document document = project.GetDocument(documentId);
                    var tree = await document.GetSyntaxTreeAsync();
                    var root = await tree.GetRootAsync();

                    var a = SyntaxFactory.Literal("Quantity Roslyn"); //the part I need to change so the token will be: someFunction(someArgument)
                    foreach (var node in root.DescendantNodes())
                    {
                        foreach (var n in node.DescendantTokens())
                        {
                            if (!CanBeMagic(n.Kind()))
                                continue;

                            if (IsWellKnownConstant(n))
                                continue;

                            string suggestion;
                            if (IsMagic(n, out suggestion))
                            {
                                if (n.Text == "\"Quantity\"")
                                {
                                    var newNode = node.ReplaceToken(n, a);
                                    var newSyntaxRoot = root.ReplaceNode(node, newNode);
                                    document = document.WithSyntaxRoot(newSyntaxRoot);
                                    project = document.Project;
                                    solution = document.Project.Solution;
                                    var abc = workspace.TryApplyChanges(solution);
                                }
                                scannedFunction(n, suggestion);
                            }
                        }
                    }
                }

So for example the code does this:

    var a = "Quantity";
    //run program, transforms into
    var a = "Quantity Roslyn";

    //what I want to do is:
    var a = "Quantity";
    //run program, transforms into
    var a = someFunction("something");

Is there any way to achieve this?

**Comment by CyrusNajmabadi** [comment]
Yes, replace the node with a new 'InvocationExpression' you create using SyntaxFactory.

**Reply by tilenzalar** [reply]
If I run this code:
`var invocationExpr = SyntaxFactory.InvocationExpression(
                                     SyntaxFactory.IdentifierName("somefunction")).WithArgumentList(
                                        SyntaxFactory.ArgumentList(
                                            SyntaxFactory.SingletonSeparatedList<ArgumentSyntax>(
                                                SyntaxFactory.Argument(
                                                    SyntaxFactory.IdentifierName("arg"))))).NormalizeWhitespace();`

I get an error at runtime in this line:
`var newSyntaxRoot = root.ReplaceNode(node, invocationExpr);`

The error:
`InvalidCastException: Unable to cast object of type 'Microsoft.CodeAnalysis.CSharp.Syntax.InvocationExpressionSyntax' to type 'Microsoft.CodeAnalysis.CSharp.Syntax.MemberDeclarationSyntax'.
`
What am I doing wrong?

**Reply by CyrusNajmabadi** [reply]
> What am I doing wrong?

you're replacing a whole member with an expression.  that's not legal.  you can replace one expression with anotehr, or a method with some other type of member.

**Reply by tilenzalar** [reply]
I'm quite new working with Roslyn. I think I get that expresion/member confusion, but I can't make it work. Do I need to create new member and add expression to it or just replace the expressions? Can't seem to find the right syntax for any of it. Is it possible for you to provide an example?
